compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/CharHelpers.skiko.kt

```diff
@@ -150,3 +152,107 @@ internal fun String.codePointBefore(index: Int): CodePoint {
     }
     return low.code
 }
+
+/**
+ * Finds the offset of the next non-whitespace symbols subsequence (word) in the given text
+ * starting from the specified caret offset.
+ *
+ * @param offset The offset where to start looking for the next word.
+ * @param currentText The current text in which to search for the next word.
+ * @return The offset of the next non-whitespace symbols subsequence (word), or the end of the string
+ *         if no such word is found.
+ */
+@ExperimentalTextApi
+fun findNextNonWhitespaceSymbolsSubsequenceStartOffset(
+    offset: Int,
+    currentText: String
+): Int {
+    /* Assume that next non whitespaces symbols subsequence (word) is when current char is whitespace and next character is not.
+     * Emoji (compound incl.) should be treated as a new word.
+     */
+    val charIterator =
+        BreakIterator.makeCharacterInstance() // wordInstance doesn't consider symbols sequence as word
+    charIterator.setText(currentText)
+
+    var currentOffset: Int
+    var nextOffset = charIterator.next()
+    while (nextOffset < offset) {
+        nextOffset = charIterator.next()
+    }
+    currentOffset = nextOffset
+
+    while (nextOffset != BreakIterator.DONE) {
+        nextOffset = charIterator.next()
+        if (currentText.codePointAt(currentOffset).isWhitespace() && !currentText.codePointAt(
+                nextOffset
+            ).isWhitespace()
+        ) {
+            return currentOffset
+        } else {
+            currentOffset = nextOffset
+        }
+    }
+    return currentOffset
+}
+
+/**
+ * Determines whether the character at the specified offset in the string is a whitespace Unicode character.
+ *
+ * @param offset The index of the character to check.
+ * @return `true` if the character at the specified offset is a whitespace character, `false` otherwise.
+ */
+@ExperimentalTextApi
+fun String.isWhitespace(offset: Int): Boolean {
+    return this.codePointAt(offset).isWhitespace()
+}
+
+/**
+ * Checks if the character at the specified offset in the string is a punctuation Unicode character.
+ *
+ * @param offset The offset of the character to check.
+ * @return true if the character at the specified offset is a punctuation character, false otherwise.
+ */
+@ExperimentalTextApi
+fun String.isPunctuation(offset: Int): Boolean {
+    return this.codePointAt(offset).isPunctuation()
+}
+
+@ExperimentalTextApi
+fun String.halfSymbolsOffset(): Int {
+    val symbolsCount = this.codePoints.count()
+    val charIterator = BreakIterator.makeCharacterInstance()
+    charIterator.setText(this)
+    var currentOffset = 0
+    for (i in 0..symbolsCount / 2) {
+        currentOffset = charIterator.next()
+    }
+    return currentOffset
+}
+
+private fun CodePoint.isWhitespace(): Boolean {
```

Comment:
Let's make it (and `isPunctuation`) internal + Kdoc

Reply:
I contacted @igordmn, we decided to move these functions to foundation due to keep all of them (including `determineCursorDesiredOffset`) internal